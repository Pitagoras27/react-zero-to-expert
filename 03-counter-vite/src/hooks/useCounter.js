import { useState } from "react";

const initialState = 21;

export const useCounter = () => {
  const [count, setCount] = useState(initialState);

  const handleAdd = () => setCount( prev => prev + 1 );

  const handleSubstract = () => setCount( prev => prev - 1 );

  const handleReset = () => setCount(initialState);

  return {
    count,
    handleAdd,
    handleSubstract,
    handleReset
  }
}