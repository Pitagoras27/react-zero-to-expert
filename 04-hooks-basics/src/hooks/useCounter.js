import { useState } from "react";

export const useCounter = (initialState = 10) => {
        const [counter, setCounter] = useState(initialState);

        /**
         * ? Is best update state with callback, because setCounter is asyncronous */
        const increment = (value = 1) => {
                setCounter((prev) => prev + value);
        };

        const reset = () => {
                setCounter(initialState);
        };
        const decrement = (value = 1) => {
                if (counter === 0) return;
                setCounter(counter - value);
        };

        return {
                counter,
                increment,
                reset,
                decrement,
        };
};
