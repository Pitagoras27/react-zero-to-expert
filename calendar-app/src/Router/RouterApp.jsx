import { Navigate, Route, Routes } from "react-router-dom";
import { LoginPage } from "../Auth";
import { CalendarPage } from "../Calendar";

export const RouterApp = () => {
  const status = 'not-authenticated';
  return (
    <Routes>
      {
        status === 'not-authenticated'
          ? <Route path='/auth/*' element={<LoginPage /> } />
          : <Route path='/*' element={<CalendarPage />} />
      }
      <Route path='/*' element={<Navigate to='/auth/login' /> } />
    </Routes>
  )
}