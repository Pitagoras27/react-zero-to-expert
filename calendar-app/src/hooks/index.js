export * from "./useCalendarStore";
export * from "./useModalForm";
export * from "./useUiStore";
