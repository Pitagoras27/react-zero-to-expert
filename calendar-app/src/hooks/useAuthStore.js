import { useDispatch, useSelector } from "react-redux";
import { calendarApi } from "../api";
import { clearErrorMessage, onChecking, onLogin, onLogout } from "../store/auth/authSlice";

export const useAuthStore = () => {
  const dispatch = useDispatch();
  const { status, errorMessage } = useSelector(state => state.auth);

  const onLoginUser = async({ email, password }) => {
    dispatch(onChecking());
    try {
      const { data } = await calendarApi.post('/auth', { email, password });
      const { name, uid, token } = data;

      localStorage.setItem('token', token);
      localStorage.setItem('token-init-date', Date.now());
      dispatch(onLogin({ name, uid }));
    } catch (error) {
        dispatch(onLogout('Incorrect credentials'));
        setTimeout(() => {
          dispatch(clearErrorMessage());
        }, 10)
      
    }
  }

  return {
    onLoginUser,
    status,
    errorMessage
  }
}