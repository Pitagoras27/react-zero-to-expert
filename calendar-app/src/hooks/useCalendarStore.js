import { useDispatch, useSelector } from "react-redux";
import calendarApi from "../api/calendarApi";
import { onAddNewEvent, onDeleteEvent, onUpdateEvent, setActiveEvent } from "../store/calendar/calendarSlice";

export const useCalendarStore = () => {
  const { events, activeEvent } = useSelector((state) => state.calendar);
  const { user } = useSelector(state => state.auth);
  const dispatch = useDispatch();

  const onSetActiveEvent = (event) => {
    dispatch(setActiveEvent(event));
  };

  const startSavingEvent = async (calendarEvent) => {
    if (calendarEvent.id) {
      dispatch(onUpdateEvent({ ...calendarEvent }));
    } else {
      const { data } = await calendarApi.post('/events', calendarEvent);
      const { id } = data?.event;
      dispatch(onAddNewEvent({ ...calendarEvent, id, user }));
    }
  };

  const startDeleteEvent = () => {
    dispatch(onDeleteEvent());
  }

  return {
    events,
    activeEvent,
    hasActiveEvent: !!activeEvent,
    onSetActiveEvent,
    startSavingEvent,
    startDeleteEvent,
  };
};
