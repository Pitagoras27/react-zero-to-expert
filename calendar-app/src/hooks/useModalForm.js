import { differenceInSeconds } from "date-fns";
import { useMemo, useRef, useState } from "react";
import Swal from "sweetalert2";

export const useModalForm = (initialState) => {
  const [formValues, setFormValues] = useState(initialState);
  const { title, notes, start, end } = formValues;
  const inputRef = useRef(null);
  const [validationText, setValidationText] = useState("");

  useMemo(() => {
    setValidationText(title.length <= 0 ? "is-invalid" : "");
  }, [validationText, title]);

  const onValueChange = ({ target }) => {
    setFormValues({
      ...formValues,
      [target.name]: target.value,
    });
  };

  const onDateSelected = (data, timeAction) => {
    setFormValues({
      ...formValues,
      [timeAction]: data,
    });
  };

  const onSubmitEvent = (event) => {
    event.preventDefault();
    const differenceTime = differenceInSeconds(end, start);

    if (differenceTime <= 0 || isNaN(differenceTime)) {
      Swal.fire({
        icon: "error",
        title: "The date is incorret",
        text: "Try again",
      });
    }
    if (title.length <= 0) inputRef.current.select();
  };

  return {
    title,
    notes,
    start,
    end,
    validationText,
    inputRef,
    setFormValues,
    onValueChange,
    onDateSelected,
    onSubmitEvent,
  };
};
