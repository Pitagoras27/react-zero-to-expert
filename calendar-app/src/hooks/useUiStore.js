import { useDispatch, useSelector } from "react-redux";
import { onModalDateClose, onModalDateOpen } from "../store";

export const useUiStore = () => {
  const { isDateModalOpen } = useSelector((state) => state.ui);
  const dispatch = useDispatch();

  const onModalOpen = () => {
    dispatch(onModalDateOpen());
  };

  const onCloseModal = () => {
    dispatch(onModalDateClose());
  };

  return {
    isDateModalOpen,
    onModalOpen,
    onCloseModal,
  };
};
