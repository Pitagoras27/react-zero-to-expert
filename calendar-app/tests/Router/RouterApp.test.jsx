import { render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { useAuthStore } from "../../src/hooks/useAuthStore";
import { RouterApp } from "../../src/Router/RouterApp";

jest.mock("../../src/hooks/useAuthStore");
jest.mock("../../src/Calendar", () => ({
    CalendarPage: () => <section>Calendar Page</section>,
  })
);

describe('test on RouterApp Component', () => {
  const mockCheckAuthToken = jest.fn();

  beforeEach(() => jest.clearAllMocks());

  test('should call checkAuthToken function and render Loading message', () => {

    useAuthStore.mockReturnValue({
      status: 'checking',
      checkAuthToken: mockCheckAuthToken
    });

    render(<RouterApp />);

    expect(mockCheckAuthToken).toHaveBeenCalled();
    expect(screen.getByText('Loading')).toBeTruthy();
  });

  test('should show register and login page', () => {
    useAuthStore.mockReturnValue({
      status: 'not-authenticated',
      checkAuthToken: mockCheckAuthToken
    });

    const { container } = render(
      <MemoryRouter initialEntries={["/auth/otherpath/login"]}>
        <RouterApp />
      </MemoryRouter>
    );

    expect(screen.getByText('Ingreso')).toBeTruthy();
    expect( container ).toMatchSnapshot();
  });

  test('should show calendar page', () => {
    useAuthStore.mockReturnValue({
      status: 'authenticated',
      checkAuthToken: mockCheckAuthToken
    });

    const { container } = render(
      <MemoryRouter>
        <RouterApp />
      </MemoryRouter>
    );

    expect(screen.getByText('Calendar Page')).toBeTruthy();
    expect( container ).toMatchSnapshot();

  })

})