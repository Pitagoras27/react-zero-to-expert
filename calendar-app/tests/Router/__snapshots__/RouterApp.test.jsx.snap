// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test on RouterApp Component should show calendar page 1`] = `
<div>
  <section>
    Calendar Page
  </section>
</div>
`;

exports[`test on RouterApp Component should show register and login page 1`] = `
<div>
  <div
    class="container login-container"
  >
    <div
      class="row"
    >
      <div
        class="col-md-6 login-form-1"
      >
        <h3>
          Ingreso
        </h3>
        <form
          aria-label="login-form"
        >
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="loginEmail"
              placeholder="Correo"
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="loginPassword"
              placeholder="Contraseña"
              type="password"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="btnSubmit"
              type="submit"
              value="Login"
            />
          </div>
        </form>
      </div>
      <div
        class="col-md-6 login-form-2"
      >
        <h3>
          Registro
        </h3>
        <form
          aria-label="register-form"
        >
          <div
            class="form-group mb-2"
          >
            <input
              aria-label="register-user-name"
              class="form-control"
              name="registerName"
              placeholder="Name"
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              aria-label="register-user-email"
              class="form-control"
              name="registerEmail"
              placeholder="Correo"
              type="email"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              aria-label="password-register"
              class="form-control"
              name="registerPassword"
              placeholder="Contraseña"
              type="password"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              aria-label="password-confirm"
              class="form-control"
              name="registerPassword2"
              placeholder="Repita la contraseña"
              type="password"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="btnSubmit"
              type="submit"
              value="Crear cuenta"
            />
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
